### DISCUSSION TITLE
Helium and arm-none-eabi-gcc version 13.2.1 and 13.3.1

### DISCUSSION THREAD

**Discussion by ilyesgouta** [ARM-software/Arm-2D#56] [General]
Hi,

I'm using arm-none-eabi-gcc version 13.2.1 20231009 (Arm GNU Toolchain 13.2.rel1 (Build arm-13.7)) to build Arm2D library through CMake and I'm getting the warning below during the build:

 81 | #           pragma GCC warning "GCC only has basic support for Helium and its\
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   82 |  performance might highly like disappoint you,You can define a macro\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |  __ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__ to suppress this warning."
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Is this warning still actual with the GCC 13.2.1 release from Arm?

If yes, then what would be the better compiler (and version) to build Arm2D?

Regards,
Ilyes

**Comment by ilyesgouta** [comment]
Hi,

I tried with the latest 13.3.rel1 release from Arm and I'm still getting the error below when building the library:

```
D:\arm-gnu-toolchain-13.3.rel1\bin\arm-none-eabi-gcc.exe --sysroot=D:/arm-gnu-toolchain-13.3.rel1/bin/../arm-none-eabi -ID:\Projects\Arm-2D\Library\Include -Wno-nonnull-compare -Wno-unused-value -fms-extensions -march=armv8.1-m.main+mve.fp+fp.dp -mfloat-abi=hard -mtune=cortex-m85 -mcmse -mthumb -fshort-enums -fno-strict-aliasing -ffunction-sections -fdata-sections -fno-exceptions -flax-vector-conversions -fomit-frame-pointer -DRTE_Acceleration_Arm_2D -DRTE_Acceleration_Arm_2D_Alpha_Blending -DRTE_Acceleration_Arm_2D_Transform -DRTE_Acceleration_Arm_2D_Filter -Wno-format -Wno-unused-function -Wno-maybe-uninitialized -MD -MT CMakeFiles/ARM2D.dir/Library/Source/arm_2d_helium.c.obj -MF CMakeFiles\ARM2D.dir\Library\Source\arm_2d_helium.c.obj.d -o CMakeFiles\ARM2D.dir\Library\Source\arm_2d_helium.c.obj -c D:\Projects\Arm-2D\Library\Source\arm_2d_helium.c
In file included from D:\Projects\Arm-2D\Library\Include/arm_2d.h:35,
                 from D:\Projects\Arm-2D\Library\Source\arm_2d_helium.c:34:
D:\Projects\Arm-2D\Library\Include/arm_2d_types.h:81:32: warning: GCC only has basic support for Helium and its performance might highly like disappoint you,You can define a macro __ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__ to suppress this warning.
   81 | #           pragma GCC warning "GCC only has basic support for Helium and its\
      |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   82 |  performance might highly like disappoint you,You can define a macro\
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   83 |  __ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__ to suppress this warning."
      |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s: Assembler messages:
C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s:141856: Error: bad element type for instruction -- `vstrb.u16 q7,[r1],#8'
CMakeFiles\ARM2D.dir\build.make:133: recipe for target 'CMakeFiles/ARM2D.dir/Library/Source/arm_2d_helium.c.obj' failed
```

**Error: bad element type for instruction -- `vstrb.u16 q7,[r1],#8**

Any thoughts on how to fix this?

Regards,
Ilyes

**Reply by GorgonMeducer** [reply]
@ilyesgouta
Yes, we continue to monitor the GCC support for Helium, and unfortunately, we think the conditions have **remained the same**.  If possible, please use [LLVM-ET](https://github.com/ARM-software/LLVM-embedded-toolchain-for-Arm/releases) instead of arm gcc.

Please define macro `__ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__` in `arm_2d_cfg.h` to suppress this warning.

For the issue you mentioned:

```
C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s: Assembler messages:
C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s:141856: Error: bad element type for instruction -- `vstrb.u16 q7,[r1],#8'
CMakeFiles\ARM2D.dir\build.make:133: recipe for target 'CMakeFiles/ARM2D.dir/Library/Source/arm_2d_helium.c.obj' failed
```

It appears that it comes from the file ccKRlDr3.s that isn't part of arm-2d.

**Reply by ilyesgouta** [reply]
Hi @GorgonMeducer,

ccKRlDr3.s is what the compiler produces when processing arm_2d_helium.c, to be passed to the assembler (gas) in order to produce the object code. Any plan to bring this directly to Arm's GCC compiler maintainers to solve it asap? It should be straightforward to fix, as the assembler will emit the equivalent armv8.1 mve opcode of that instruction. This issue has been around for one year already.

**Reply by ilyesgouta** [reply]
Btw @GorgonMeducer, and for the ease of use, do you have the LLVM ET toolset file for CMake to quickly build the lib? If so, would it be possible to push it to the Arm2D repository?

**Reply by GorgonMeducer** [reply]
> It should be straightforward to fix, as the assembler will emit the equivalent armv8.1 mve opcode of that instruction. This issue has been around for one year already.

Sorry for this inconvenience. I am afraid there is no guarantee that any helium-related issue will be fixed in GCC in the near term.  In fact, the focus has moved away from GCC to LLVM-ET for many years. 
The conclusion is that for anything involving Helium, we highly recommend people use LLVM-ET or Arm Compiler 6 (also based on LLVM). 

> do you have the LLVM ET toolset file for CMake to quickly build the lib? 

I am sorry, we only have a CMakeLists [here](https://github.com/ARM-software/Arm-2D/blob/main/CMakeLists.txt).  
Our development team is small. We use [PC project (Makefile-based)](https://github.com/ARM-software/Arm-2D/tree/main/examples/%5Btemplate%5D%5Bpc%5D%5Bvscode%5D) and MDK (Arm Compiler 6) most of the time.  We rely on an open-source community to maintain cmake based building system (#42)

**Reply by adam9500370** [reply]
> Hi,
> 
> I tried with the latest 13.3.rel1 release from Arm and I'm still getting the error below when building the library:
> 
> ```
> D:\arm-gnu-toolchain-13.3.rel1\bin\arm-none-eabi-gcc.exe --sysroot=D:/arm-gnu-toolchain-13.3.rel1/bin/../arm-none-eabi -ID:\Projects\Arm-2D\Library\Include -Wno-nonnull-compare -Wno-unused-value -fms-extensions -march=armv8.1-m.main+mve.fp+fp.dp -mfloat-abi=hard -mtune=cortex-m85 -mcmse -mthumb -fshort-enums -fno-strict-aliasing -ffunction-sections -fdata-sections -fno-exceptions -flax-vector-conversions -fomit-frame-pointer -DRTE_Acceleration_Arm_2D -DRTE_Acceleration_Arm_2D_Alpha_Blending -DRTE_Acceleration_Arm_2D_Transform -DRTE_Acceleration_Arm_2D_Filter -Wno-format -Wno-unused-function -Wno-maybe-uninitialized -MD -MT CMakeFiles/ARM2D.dir/Library/Source/arm_2d_helium.c.obj -MF CMakeFiles\ARM2D.dir\Library\Source\arm_2d_helium.c.obj.d -o CMakeFiles\ARM2D.dir\Library\Source\arm_2d_helium.c.obj -c D:\Projects\Arm-2D\Library\Source\arm_2d_helium.c
> In file included from D:\Projects\Arm-2D\Library\Include/arm_2d.h:35,
>                  from D:\Projects\Arm-2D\Library\Source\arm_2d_helium.c:34:
> D:\Projects\Arm-2D\Library\Include/arm_2d_types.h:81:32: warning: GCC only has basic support for Helium and its performance might highly like disappoint you,You can define a macro __ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__ to suppress this warning.
>    81 | #           pragma GCC warning "GCC only has basic support for Helium and its\
>       |                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>    82 |  performance might highly like disappoint you,You can define a macro\
>       |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>    83 |  __ARM_2D_SUPPRESS_GCC_HELIUM_PERFORMANCE_WARNING__ to suppress this warning."
>       |  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s: Assembler messages:
> C:\Users\goutai\AppData\Local\Temp\ccKRlDr3.s:141856: Error: bad element type for instruction -- `vstrb.u16 q7,[r1],#8'
> CMakeFiles\ARM2D.dir\build.make:133: recipe for target 'CMakeFiles/ARM2D.dir/Library/Source/arm_2d_helium.c.obj' failed
> ```
> 
> **Error: bad element type for instruction -- `vstrb.u16 q7,[r1],#8**
> 
> Any thoughts on how to fix this?
> 
> Regards, Ilyes

I also encountered this error when building with arm gcc. It is caused by [this typo](https://github.com/ARM-software/Arm-2D/blob/developing/Library/Source/__arm_2d_tile_copy_with_source_mask_and_opacity_helium.c#L720).
I created the [pull request](https://github.com/ARM-software/Arm-2D/pull/57) to fix.

**Reply by GorgonMeducer** [reply]
I didn't expect a typo could be the reason behind it. Thank you very much.

**Reply by ilyesgouta** [reply]
Hi,

Good finding @adam9500370. Has this been merged?

Has Arm-2D been (extensively) tested using GCC intrinsics w.r.t performance? any testing report in the repository?

Thank you,

Ilyes

**Reply by GorgonMeducer** [reply]
@ilyesgouta Merged. 
There is a PC project template in the example folder, and this PC project is using GCC.

**Reply by GorgonMeducer** [reply]
We still highly recommend people use LLVM-ET rather than GCC.
